From 4407a55002ca8a265ad611dd0f563ddab4a01947 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:09:28 -0400
Subject: [virt] kvm/ppc: book3s hv - Flush the correct number of TLB sets on POWER8

Message-id: <1410545655-205645-340-git-send-email-dzickus@redhat.com>
Patchwork-id: 94373
O-Subject: [RHEL7 PATCH 339/626] KVM: PPC: Book3S HV: Flush the correct number of TLB sets on POWER8
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit ca252055130b6a1affa12df94a4694c1aafc2a6c
Author: Paul Mackerras <paulus@samba.org>
Date:   Wed Jan 8 21:25:22 2014 +1100

    KVM: PPC: Book3S HV: Flush the correct number of TLB sets on POWER8

    POWER8 has 512 sets in the TLB, compared to 128 for POWER7, so we need
    to do more tlbiel instructions when flushing the TLB on POWER8.

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv_rmhandlers.S b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
index 804de0e..b9bcd04 100644
--- a/arch/powerpc/kvm/book3s_hv_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
@@ -434,7 +434,13 @@ END_FTR_SECTION_IFSET(CPU_FTR_ARCH_201)
  andc r7,r7,r0
  stdcx. r7,0,r6
  bne 23b
- li r6,128   /* and flush the TLB */
+ /* Flush the TLB of any entries for this LPID */
+ /* use arch 2.07S as a proxy for POWER8 */
+BEGIN_FTR_SECTION
+ li r6,512   /* POWER8 has 512 sets */
+FTR_SECTION_ELSE
+ li r6,128   /* POWER7 has 128 sets */
+ALT_FTR_SECTION_END_IFSET(CPU_FTR_ARCH_207S)
  mtctr r6
  li r7,0x800  /* IS field = 0b10 */
  ptesync
-- 
1.7.1